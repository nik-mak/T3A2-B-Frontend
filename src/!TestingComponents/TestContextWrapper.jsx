import React from 'react'
import AlertsContext from "../contexts/alert";
import CartContext from "../contexts/cart";
import ModalsContext from "../contexts/modals";
import SelectedListingContext from "../contexts/selectedListing";
import CartTotalContext from "../contexts/total";
import UserContext from "../contexts/user";

function TestContextWrapper( {children, user} ) {

  return (
    <SelectedListingContext.Provider value={[]}>
          <ModalsContext.Provider value={[]}>
            <AlertsContext.Provider value={[]}>
              <CartContext.Provider value={[]}>
                <CartTotalContext.Provider value={[]}>
                  <UserContext.Provider
                    value={[user]}
                  >
                    {children}
                  </UserContext.Provider>
                </CartTotalContext.Provider>
              </CartContext.Provider>
            </AlertsContext.Provider>
          </ModalsContext.Provider>
        </SelectedListingContext.Provider>
  )
}

export default TestContextWrapper