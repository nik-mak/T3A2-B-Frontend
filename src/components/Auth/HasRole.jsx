import React, { useContext } from "react";
import UserContext from "../../contexts/user";

/**
 * HasRole component is used to render children components if a user is logged in and has a role included in the roles param
 *
 * @param {Array} roles an array of the roles permitted 
 * @param {HTML} children components to be rendered if authorization is successful
 * @return {HTML} children components or nothing depending on authorization success
 */
function HasRole({ roles, children }) {
  const user = useContext(UserContext)[0];

  return <>{user && roles.includes(user.role) && (children)}</>;
}

export default HasRole;
