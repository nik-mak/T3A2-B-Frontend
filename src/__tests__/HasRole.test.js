import renderer from "react-test-renderer";
import TestContextWrapper from "../!TestingComponents/TestContextWrapper";
import HasRole from "../components/auth/HasRole";

it("renders when the user role matches a role in the roles prop", () => {
  const component = renderer.create(
    <TestContextWrapper
      user={{ name: "Test", email: "test@test", role: "staff" }}
    >
      <HasRole roles={["staff"]}>TEST</HasRole>
    </TestContextWrapper>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("doesn't render when the user role doesn't match a role in the roles prop", () => {
  const component = renderer.create(
    <TestContextWrapper
      user={{ name: "Test", email: "test@test", role: "admin" }}
    >
      <HasRole roles={["staff"]}>TEST</HasRole>
    </TestContextWrapper>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("doesn't render when the user role doesn't match a role in an empty roles prop", () => {
  const component = renderer.create(
    <TestContextWrapper
      user={{ name: "Test", email: "test@test", role: "admin" }}
    >
      <HasRole roles={[]}>TEST</HasRole>
    </TestContextWrapper>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("doesn't render when the user is undefined", () => {
  const component = renderer.create(
    <TestContextWrapper user={undefined}>
      <HasRole roles={["admin"]}>TEST</HasRole>
    </TestContextWrapper>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders when the user role full roles prop", () => {
  const component = renderer.create(
    <TestContextWrapper user={undefined}>
      <HasRole roles={["admin", "customer", "dragon"]}>TEST</HasRole>
    </TestContextWrapper>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
