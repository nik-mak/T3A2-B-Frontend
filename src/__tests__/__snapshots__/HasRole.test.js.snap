// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't render when the user is undefined 1`] = `null`;

exports[`doesn't render when the user role doesn't match a role in an empty roles prop 1`] = `null`;

exports[`doesn't render when the user role doesn't match a role in an empty roles prop 2`] = `null`;

exports[`doesn't render when the user role doesn't match a role in the roles prop 1`] = `null`;

exports[`renders when the user role full roles prop 1`] = `null`;

exports[`renders when the user role matches a role in the roles prop 1`] = `"TEST"`;
